packages/adapter-knex/lib/adapter-knex.js

```diff
@@ -277,35 +292,30 @@ class KnexListAdapter extends BaseListAdapter {
       .del();
   }
 
-  async _populateMany(result) {
-    // Takes an existing result and merges in all the many-relationship fields
-    // by performing a query on their join-tables.
-    return {
-      ...result,
-      ...(await resolveAllKeys(
-        arrayToObject(
-          this.fieldAdapters.filter(a => a.isRelationship && a.config.many),
-          'path',
-          async a =>
-            (await this._query()
-              .select(a.refListId)
-              .from(this._manyTable(a.path))
-              .where(`${this.key}_id`, result.id)
-              .returning(a.refListId)).map(x => x[a.refListId])
-        )
-      )),
-    };
+  // Add subqueries that return arrays of IDs for each "many" relationship
+  _addManySubqueries(query, baseTableAlias) {
+    const manyAdapters = this.fieldAdapters.filter(a => a.isRelationship && a.config.many);
+    query.column(manyAdapters.map(a => {
+      const manyTable = this._manyTable(a.path);
+      const manyTableAlias = `s${baseTableAlias}__${a.path}`;
+      return this._query().select(this._knexRaw('array_agg(??)', `${manyTableAlias}.${a.refListId}`))
```

Comment:
Adding `array_agg` will tie our adapter directly to postgres. We want to retain as much support for different DBs via knex as possible (hence why we chose knex, and not postgres directly). Is there another way to do this without dropping to raw SQL?

Reply:
No, array operations aren't supported by Knex.

Unfortunately, array handling isn't very portable.  (I didn't realise KS is meant to support DBs other than PG.)  But doing something like this is a lot better for performance than looping over the results and running extra queries.  Maybe it's worth special-casing for PG and MySQL, with a fallback to looping.